[WIP] WebSocket Tree Map and Payload Analyzer

eakirtas opened the pull request:
This pull request its part of WebSocket Active Scan project which described in issue [#4276](https://github.com/zaproxy/zaproxy/issues/4276).

_______________________________________________________
- [ ]  **[WS-Tree-Map] WebSocket Tree Map Structure**

Introducing to new WebSocket Tree:

- Nodes: All WebSocket Nodes should implement the interface
  StructuralWebSocketNode. That interface helps to store the tree structure
  on a database in the future.

   - `WebSocketTreeNode` The base class of all Node, it is an abstract class
     which offers the basic method to create the nodes and navigate over them.

  - `WebSocketNodeType`: Helps to define the different use of the node. The
     different types maybe will act differently. (ex. the `FOLDER_HANDSHAKES`
     keeps the handshakes of a connection, `MESSAGE_PONG` it is a node which
     keeps ping messages).

  - `WebSocketFolderNode`: These nodes act as a node folder which keeps a
     specific kind of child. (ex `FOLDER_HOST` always keeps the appropriate
     folders for a host, `FOLDER_HANDSHAKES` keeps the paths of the handshakes)

  - `WebSocketMessageNode`: Keeps the WebSocket Messages and the appropriate
     stuff to analyze the messages.  WebSocket Messages represented as DTOs.

  - `WebSocketHandshakeNode`: Keeps the HTTP handshake.

  All channels have the same following structure:

  - ROOT_FOLDER
      - HOST_FOLDER
         - HANDSHAKE_FOLDER
            - _path_ HANDSHAKE_FOLDER
               - _to_ HANDSHAKE_FOLDER
                  - _Handshake_ HANDSHAKE
         - HEARTBEAT_FOLDER
            - _Ping Message_ PING_MESSAGE
            - _Pong Message_ PING_MESSAGE
         - MESSAGES_FOLDER
            - _Analyzed Payload Structure_ MESSAGE
         - CLOSE_FOLDER
            - _Close Message_ CLOSE_MESSAGE

 * `WebSocketMap`:
    Manage the whole functionality of the tree. Add connections and Messages to appropriate position in the tree. Initialize a new tree by the static method createTree.

 TODO:
  - [x] Correct the following: path to the handshake represented as Handshake type not as HandshakeFolders
  - [x] Implement UI
  - [ ] Implement API
  - [ ] Add documentation
  - [ ] Extend the existing interface StructuralWebSocketNode
  - [x] Sort folders under HOST_FOLDER
  - [ ] Remove Changes related with whitespaces 
  - [ ] Add script mechanism

_______________________________________________________
- [ ]  **[WS-Tree-Map] Tests for node and the map** 

 TODO: 
- [ ] Add more test cases
_______________________________________________________
- [ ]  **[WS-Tree-Map] Payload Analyzer Structure and JSON analyzer**

This commit aims to introduce a formal method to analyze and "represent"
WebSocket message payload. Every analyzer should implement the `PayloadAnalyzer` class
and produce a `PayloadStructure`. Every payload should be analyzed into
`PayloadStructure`.

Json analyzer uses the [GSON](https://github.com/google/gson) library to parse the payloads. The first analyzer called `LazyJsonAnalyzer` which serialize and deserialize the messages to/from `LazyJsonStructure`.  `LazyJsonStructure` maps the JSON message into a HashMap which keeps the keys, type of values and the values. That class use recursive methods to analyze the message. As a result, create a tree similar to JSON structure.  

TODO:
- [ ] Implement an API
- [ ] Add html Analyzer.
- [ ] Add methods in order to iterate on diferent paremeters
_______________________________________________________
- [ ]  **[WS-Tree-Map] Test For Payload Analyzer** 
_______________________________________________________

- [ ] **[WS-Tree-Map] Basic UI**

User interface for WebSocket Tree

TODO: 
- [ ] Selecting the Map node reflected in the Req/Resp tabs.
- [ ] Address the Rendering Issues
_______________________________________________________

**Web Application could be tested** (tick if works fine)

- [X] http://mod-synth.io/
- [ ] http://browserquest.mozilla.org/

eakirtas committed:
[WS-Tree-Map] WebSocket Tree Map Structure

Introducing to new WebSocket Tree:

* Nodes: All WebSocket Nodes should implement the interface
  StructuralWebSocketNode. That interface helps to store the tree structure
  on a database in the future.

  ** WebSocketTreeNode The base class of all Node, it is an abstract class
     which offers the basic method to create the nodes and navigate over them.

  ** WebSocketNodeType: Helps to define the different use of the node. The
     different types maybe will act differently. (ex. the FOLDER_HANDSHAKES
     keeps the handshakes of a connection, MESSAGE_PONG it is a node which
     keeps ping messages).

  ** WebSocketFolderNode: This nodes acts as a node folder which keeps a
     specific kind of child. (ex FOLDER_HOST always keeps the appropriate
     folders for a host, FOLDER_HANDSHAKES keeps the paths of the handshakes)

  ** WebSocketMessageNode: Keeps the WebSocket Messages and the appropriate
     stuff to analyze the messages.  WebSocket Messages represented as DTOs.

  ** WebSocketHandshakeNode: Keeps the http handshake.

  All channels has the same following structure:

  |- ROOT_FOLDER
      |- HOST_FOLDER
         |- HANDSHAKE_FOLDER
            |- <path> HANDSHAKE_FOLDER
               |- <to> HANDSHAKE_FOLDER
                  |- <Handshake> HANDSHAKE
         |- HEARTBEAT_FOLDER
            |- <Ping Message> PING_MESSAGE
            |- <Pong Message> PING_MESSAGE
         |- MESSAGES_FOLDER
            |- (Analyzed Payload Structure) MESSAGE
         |- CLOSE_FOLDER
            |- <Close Message> CLOSE_MESSAGE

* WebSocketMap:
  Manage whole functionality of the tree. Add connections and Messages to
  appropriate position in the tree. Initialize a new tree by static method
  createTree.

TODO:
 [ ] Correct the following: path to the handshake represented as
  Handshake type not as HandshakeFolders
 [ ] Implement UI
 [ ] Implement API
 [ ] Add documentation
 [ ] Extend the existing interface StructuralWebSocketNode
 [ ] Sort folders under HOST_FOLDER

eakirtas committed:
[WS-Tree-Map] Tests for node and the map

eakirtas committed:
[WS-Tree-Map] Test For Payload Analyzer

eakirtas committed:
[WS-Tree-Map] Basic UI

eakirtas committed:
[WS-Tree-Map] Some changes to Adapt Structural Tree to UI

[WS-Tree-Map] Add History references & add some constants

psiinon commented:
Doesnt seem to work so well with http://browserquest.mozilla.org/ - this has a load of messages like:
[1,53578907,"psi",16,234,80]
[[17,5,null],[19,929,931,1220,1221,1222,1320,1321,1322,1720,1721,11920,11921,12020,12021,12120,12121,810235,815222,815246,53278903]]
[20,929,931,1220,1221,1222,1320,1321,1322,1720,1721,11920,11921,12020,12021,12120,12121,810235,815222,815246,53278903]
[[2,929,61,19,233],[2,931,61,53,249],[2,1220,2,47,223,2],[2,1221,2,46,223,3],[2,1222,2,47,223,4],[2,1320,2,9,228,4],[2,1321,2,21,223,1],[2,1322,2,21,225,3],[2,1720,2,10,245,2],[2,1721,2,14,243,1],[2,11920,2,36,233,2],[2,11921,2,39,237,4],[2,12020,2,50,246,2],[2,12021,2,50,245,1],[2,12120,2,24,235,3],[2,12121,2,24,235,1],[2,810235,42,10,235],[2,815222,43,15,222],[2,815246,50,15,246],[2,53278903,1,15,235,"fff",1,21,60]]

Would also really like to see selecting the Map node reflected in the Req/Resp tabs.
Also, the button doesnt  seem to work, is is suppose to?

But I do really like seeing this start to take shape :)

psiinon commented:
This pull request **introduces 9 alerts** when merging 9ac6d5b8b986effdece8b24db837f56f7a1bffaf into b33ac46d6bb7db10ed20d53a51ff48dd8d3010de - [view on LGTM.com](https://lgtm.com/projects/g/zaproxy/zap-extensions/rev/pr-ebc31b4a64bbeee542afaab23254184f53ee137d)

**new alerts:**

* 2 for Container contents are never accessed
* 2 for Equals on incomparable types
* 2 for Implicit conversion from array to string
* 1 for Inconsistent equals and hashCode
* 1 for Reference equality test of boxed types
* 1 for Dereferenced variable may be null

---

*Comment posted by [LGTM.com](https://lgtm.com)*

psiinon commented:
This pull request **introduces 8 alerts** when merging 7b8cb57ea05399b4463fe9d40b1ea1bcbe7d4f9f into b33ac46d6bb7db10ed20d53a51ff48dd8d3010de - [view on LGTM.com](https://lgtm.com/projects/g/zaproxy/zap-extensions/rev/pr-b0bbb429f6d584839a751e0d8054cfcd15235211)

**new alerts:**

* 2 for Equals on incomparable types
* 2 for Implicit conversion from array to string
* 1 for Container contents are never accessed
* 1 for Inconsistent equals and hashCode
* 1 for Reference equality test of boxed types
* 1 for Dereferenced variable may be null

---

*Comment posted by [LGTM.com](https://lgtm.com)*

psiinon commented:
This pull request **introduces 8 alerts** when merging 500e31b860fcfbd9ae453d5e8fa7842245f2e547 into b33ac46d6bb7db10ed20d53a51ff48dd8d3010de - [view on LGTM.com](https://lgtm.com/projects/g/zaproxy/zap-extensions/rev/pr-18a2701e1a3a3f05d17bf6eae5600cecd666e8c0)

**new alerts:**

* 2 for Equals on incomparable types
* 2 for Implicit conversion from array to string
* 1 for Container contents are never accessed
* 1 for Inconsistent equals and hashCode
* 1 for Reference equality test of boxed types
* 1 for Dereferenced variable may be null

---

*Comment posted by [LGTM.com](https://lgtm.com)*

ManosMagnus commented:
@psiinon  The button just added in order to test the layout. This button will be used on future to establish a new connection. The suggestion added to the todo list and will be addressed as soon as possible.

psiinon commented:
This pull request **introduces 8 alerts** when merging 500e31b860fcfbd9ae453d5e8fa7842245f2e547 into de4537e305d018e8a185c0b7feb5f87576f021c9 - [view on LGTM.com](https://lgtm.com/projects/g/zaproxy/zap-extensions/rev/pr-b0677d2d099e36119d57710e7d70ab9bb589311a)

**new alerts:**

* 2 for Equals on incomparable types
* 2 for Implicit conversion from array to string
* 1 for Container contents are never accessed
* 1 for Inconsistent equals and hashCode
* 1 for Reference equality test of boxed types
* 1 for Dereferenced variable may be null

---

*Comment posted by [LGTM.com](https://lgtm.com)*

kingthorin commented:
@ManosMagnus are you able to address any of the outstanding todo (checkbox) items or LGTM findings?

ManosMagnus commented:
yeah, I'll be back next week to work on my PRs :)

kingthorin commented:
@ManosMagnus how are things going?

ManosMagnus commented:
According to the latest JSON specification, the following syntax is valid
```
123
"string"
null
true
false
["array item 1", 123]
```
LazyJsonAnalyzer is updated accordingly. So browserquest's examples seems to be handled right. Sorry but I can't test it with http://browserquest.mozilla.org/, website doesnt work correctly for me.

psiinon commented:
This pull request **introduces 7 alerts** when merging b905e899d90439fd5cba9187bb4c78a4d8ff6101 into 7cd4f60afe1bd3294aad3251c36e66330043642b - [view on LGTM.com](https://lgtm.com/projects/g/zaproxy/zap-extensions/rev/pr-a80cd3f6b6a22bbb2b83a5b9c0a06cf27e8695a4)

**new alerts:**

* 2 for Implicit conversion from array to string
* 1 for Container contents are never accessed
* 1 for Equals on incomparable types
* 1 for Inconsistent equals and hashCode
* 1 for Reference equality test of boxed types
* 1 for Dereferenced variable may be null

---

*Comment posted by [LGTM.com](https://lgtm.com)*

psiinon commented:
This pull request **introduces 3 alerts** when merging f1dd88adbd99fed3182e327b00e272eb3ce32a46 into 0ceff13117188accffd473b5c36075b6d79ef55a - [view on LGTM.com](https://lgtm.com/projects/g/zaproxy/zap-extensions/rev/pr-4d7c56c6b8bdf602e545798bc8dc9b6000e11923)

**new alerts:**

* 1 for Container contents are never accessed
* 1 for Equals on incomparable types
* 1 for Expression always evaluates to the same value

---

*Comment posted by [LGTM.com](https://lgtm.com)*

thc202 commented:
The pull request is now updated (after build changes, zaproxy/zaproxy#5302).

ManosMagnus commented:
Thank you @thc202

thc202 commented:
Restored the commit lost.

psiinon commented:
This pull request **introduces 3 alerts** when merging 917bb6e029df90735caf73bee8fe9de1b61f2c2b into ffcb98d4a2565fffa21227ff78c3537f35d844ed - [view on LGTM.com](https://lgtm.com/projects/g/zaproxy/zap-extensions/rev/pr-9102f4037c37fe9c368a8a2bf17fbfd7dea717bd)

**new alerts:**

* 1 for Container contents are never accessed
* 1 for Equals on incomparable types
* 1 for Expression always evaluates to the same value

---

*Comment posted by [LGTM.com](https://lgtm.com)*

eakirtas committed:
Websocket Gradle: Add GSON library

eakirtas committed:
[WS-Tree-Map] Adapt WebSocket Tree Map Structure to Payload Analyzer

eakirtas committed:
[WS-Tree-Map] Payload Analyzer Structure and JSON analyzer.

This commit aims to introduce a formal method to analyze and "represent"
websocket message payload. Every analyzer should implement the PayloadAnalyzer class
and produce a PayloadStructure. Every payload should be analyzed into
PayloadStructure.

TODO:
  [ ] Implement an API
  [ ] Add html Analyzer.
  [ ] Add methods in order to iterate on diferent paremeters

ManosMagnus commented:
Tweak commits

psiinon commented:
Thank you for this PR, and sorry it has taken us so long to address it.
We have decided to close it, but if you are interested in revisiting it, or in contributing to ZAP in any way then please get in touch.
